### PR TITLE
Add resource filter for ARM Extension API requests

### PR DESCRIPTION
**Pull request by mathewc**
DO NOT MERGE (waiting for partner deployments). Corresponding v1 PR is [here](https://github.com/Azure/azure-functions-host/pull/5351).

### CODE REVIEW

File: src/WebJobs.Script.WebHost/Filters/ArmExtensionResourceFilter.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    public sealed class ArmExtensionResourceFilter : IAsyncResourceFilter
+    {
+        public async Task OnResourceExecutionAsync(ResourceExecutingContext context, ResourceExecutionDelegate next)
+        {
+            bool isArmExtensionRequest = !string.IsNullOrEmpty(context.HttpContext.Request.GetHeaderValueOrDefault(ScriptConstants.AntaresARMExtensionsRouteHeader));
+            bool isOwnerOrCoAdmin = string.Equals(context.HttpContext.Request.GetHeaderValueOrDefault(ScriptConstants.AntaresClientAuthorizationSourceHeader), "legacy", StringComparison.OrdinalIgnoreCase);
+            bool isRoleBasedContributor = string.Equals(context.HttpContext.Request.GetHeaderValueOrDefault(ScriptConstants.AntaresRoleBasedContributorHeader), "1");
+
+            if (isArmExtensionRequest && string.Equals(context.HttpContext.Request.Method, "GET", StringComparison.OrdinalIgnoreCase) &&
+                !isOwnerOrCoAdmin && !isRoleBasedContributor)
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Filters/ArmExtensionResourceFilter.cs]: These are the same Owner/RBAC checks that are already performed in Kudu for similar purposes [here](https://github.com/projectkudu/kudu/blob/e57833fea88995d6a24c8c06f956cc291838bdc1/Kudu.Services.Web/Tracing/TraceModule.cs#L74)

File: src/WebJobs.Script.WebHost/Filters/ArmExtensionResourceFilter.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    /// <summary>
+    /// Resource filter used to ensure secrets aren't returned for GET requests made via the Functions ARM extension
+    /// API (hostruntime), unless properly authorized.
+    /// </summary>
+    public sealed class ArmExtensionResourceFilter : IAsyncResourceFilter
+    {
+        public async Task OnResourceExecutionAsync(ResourceExecutingContext context, ResourceExecutionDelegate next)
+        {
+            bool isArmExtensionRequest = !string.IsNullOrEmpty(context.HttpContext.Request.GetHeaderValueOrDefault(ScriptConstants.AntaresARMExtensionsRouteHeader));
+            bool isOwnerOrCoAdmin = string.Equals(context.HttpContext.Request.GetHeaderValueOrDefault(ScriptConstants.AntaresClientAuthorizationSourceHeader), "legacy", StringComparison.OrdinalIgnoreCase);
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Filters/ArmExtensionResourceFilter.cs]: When running via ArmClient, I see this "Legacy" header is set

File: src/WebJobs.Script.WebHost/Filters/ArmExtensionResourceFilter.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+            bool isRoleBasedContributor = string.Equals(context.HttpContext.Request.GetHeaderValueOrDefault(ScriptConstants.AntaresRoleBasedContributorHeader), "1");
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Filters/ArmExtensionResourceFilter.cs]: @suwatch this header will only be set to 1 if the request has been properly RBAC authorized? When testing via ArmClient I don't see this header set.

File: src/WebJobs.Script.WebHost/Filters/ArmExtensionResourceFilter.cs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            bool isArmExtensionRequest = request.HasHeader(ScriptConstants.AntaresARMRequestTrackingIdHeader) &&
+                request.HasHeader(ScriptConstants.AntaresARMExtensionsRouteHeader);
+
+            if (isArmExtensionRequest && string.Equals(request.Method, "GET", StringComparison.OrdinalIgnoreCase))
+            {
+                // requests made by owner/co-admin are not filtered
+                if (!request.HasHeaderValue(ScriptConstants.AntaresClientAuthorizationSourceHeader, "legacy"))
+                {
+                    var controllerActionDescriptor = context.ActionDescriptor as ControllerActionDescriptor;
+                    if (controllerActionDescriptor != null && controllerActionDescriptor.MethodInfo != null &&
```
Comment (paulbatum) [inline src/WebJobs.Script.WebHost/Filters/ArmExtensionResourceFilter.cs]: in what scenario can these two null checks fail?

Comment (mathewc) [reply src/WebJobs.Script.WebHost/Filters/ArmExtensionResourceFilter.cs]: None really - I was just being defensive here. A resource filter will only run after route matching has selected an action, and all authz has happened. So we'll have resolved to an action at this point.

**Pull request merged**